### ISSUE TITLE
Яндекс.Модуль постоянно меняет статус на unavailable и обратно

### ISSUE THREAD

**Issue by dskudrin** [AlexxIT/YandexStation#568]
Добрый день.
Локально подключены 4 штуки Я.Модуль-2. Каждые 1-2 минуты отваливаются и возвращаются обратно.

```
Эта ошибка возникла в пользовательской интеграции.

Регистратор: custom_components.yandex_station.core.yandex_glagol
Источник: custom_components/yandex_station/core/yandex_glagol.py:152
интеграция: Yandex.Station (документация, проблемы)
Первое сообщение: 12:14:39 (77 сообщений)
Последнее сообщение: 14:20:49

Модуль гостиная => local | Exception('https://quasar.yandex.net/glagol/token return 400 status')
Модуль кухня => local | Exception('https://quasar.yandex.net/glagol/token return 400 status')
Модуль спальня => local | Exception('https://quasar.yandex.net/glagol/token return 400 status')
Модуль гостевая => local | Exception('https://quasar.yandex.net/glagol/token return 400 status')
Модуль кухня => local | AttributeError("'NoneType' object has no attribute '_timeout_ceil_threshold'")
```

**Comment by AlexxIT** [comment]
Что-то с авторизацией. А какие-то устройства вообще работают через эту интеграцию?

**Comment by dskudrin** [comment]
Подключены еще станции мини-2, всего 12 устройств вместе с модулями. Сейчас посмотрел - станции тоже периодически отваливаются, но гораздо реже - раз в 1-2 часа

**Comment by AlexxIT** [comment]
Возможно проблемы или с WiFi или с Интернетом.

**Comment by dskudrin** [comment]
Ни у каких других устройств таких проблем нет, в том числе и яндексовских. Только у модулей, и у всех одинаковые симптомы.

**Comment by AlexxIT** [comment]
Сложно сказать. Нужно ещё найти пользователей с такой же проблемой. Чтоб понять, массовая она или нет.

**Comment by dskudrin** [comment]
Проблема всё ещё актуальна на 24.08.2026.

`Яндекс.Модуль` по-прежнему периодически переходит в `unavailable`, после чего самостоятельно возвращается в нормальное состояние. Сегодня, например, по истории HA были такие интервалы:

* 16:54:29 → 16:55:59 — около 1,5 мин
* 16:56:10 → 17:58:57 — около 63 мин
* 18:54:00 → 18:59:01 — около 5 мин

То есть это уже не только кратковременные переподключения на несколько секунд.

Вижу также похожий issue #719, где фигурируют ошибки `glagol/token return 400` и `429`, поэтому интересно, связана ли текущая проблема Модулей с тем же механизмом получения токена локального режима.

Если нужно, могу включить debug для

`custom_components.yandex_station.core.yandex_glagol`

и собрать лог непосредственно в момент перехода Модуля в `unavailable`.

Есть ли сейчас понимание, в какую сторону лучше копать, или какие данные/логи будут наиболее полезны для диагностики?

**Comment by AlexxIT** [comment]
Нет, понимания нет. Только если вы самостоятельно разберётесь с проблемой.

**Comment by dskudrin** [comment]
Похоже, удалось локализовать причину и подтвердить её локальным фиксом.

У нескольких Yandex Module 2 mDNS/SRV server одинаковый — Android.local.

Из-за этого python-zeroconf объединяет A-записи разных физических Модулей, и для каждого отдельного deviceId info.addresses начинает содержать IP нескольких устройств.

Например, одновременно получаю:

deviceId=<module1> server=Android.local addresses=['192.168.3.55', '192.168.3.54', '192.168.3.52']
deviceId=<module2> server=Android.local addresses=['192.168.3.55', '192.168.3.54', '192.168.3.52']
deviceId=<module3> server=Android.local addresses=['192.168.3.55', '192.168.3.54', '192.168.3.52']

При этом порядок адресов периодически меняется.

Сейчас в YandexIOListener._zeroconf_handler() используется:

"host": str(ipaddress.ip_address(info.addresses[0])),

Поэтому при изменении порядка addresses интеграция считает, что IP устройства изменился. Далее start_or_restart() закрывает текущий WebSocket, после чего сущность уходит в unavailable, заново запрашивается glagol token и начинается reconnect.

Для проверки я сделал локальный workaround: перед add_handler() жёстко сопоставил deviceId -> реальный IP для четырёх своих Модулей.

После этого Zeroconf по-прежнему отдаёт смешанные списки IP, но:

Обновление IP-адреса устройства — 0
Возврат в облачный режим — 0
Ошибка подключения — 0
каждый Модуль сделал только одно локальное подключение после запуска HA
локальные команды продолжают работать

До фикса переключения IP происходили постоянно, иногда несколько раз в минуту.

Похоже, здесь нужно не использовать безусловно info.addresses[0]. Как минимум, если для данного deviceId уже есть рабочий host и он всё ещё присутствует в info.addresses, имеет смысл сохранять текущий адрес и не закрывать активный WebSocket только из-за изменения порядка списка.

Если нужно, могу приложить полный debug-log и свой временный патч.

**Comment by AlexxIT** [comment]
Можете показать все поля, которые прилетают в zeroconf_handler от двух разных устройств?

**Comment by dskudrin** [comment]
Собрал полный ServiceInfo. Получилась довольно показательная картина.

Два разных физических Module 2 практически одновременно:
```
deviceId='D00B4B300TNDEW'
platform='yandexmodule_2'
server='Android.local.'
port=1961
weight=0
priority=0
addresses=['192.168.3.55', '192.168.3.52', '192.168.3.54']
parsed_addresses=[
  '192.168.3.55',
  '192.168.3.52',
  '192.168.3.54',
  'fe80::ba87:6eff:fe47:7fce',
  'fe80::ba87:6eff:fe47:4ff8',
  'fe80::ba87:6eff:fe47:2754'
]
properties={
  'td': 'follower',
  'cluster': 'yes',
  'deviceId': 'D00B4B300TNDEW',
  'platform': 'yandexmodule_2'
}
interface_index=None
host_ttl=120
other_ttl=4500
```
```
deviceId='D00KSA300VJDWW'
platform='yandexmodule_2'
server='Android.local.'
port=1961
weight=0
priority=0
addresses=['192.168.3.55', '192.168.3.52', '192.168.3.54']
parsed_addresses=[
  '192.168.3.55',
  '192.168.3.52',
  '192.168.3.54',
  'fe80::ba87:6eff:fe47:7fce',
  'fe80::ba87:6eff:fe47:4ff8',
  'fe80::ba87:6eff:fe47:2754'
]
properties={
  'td': 'follower',
  'cluster': 'yes',
  'deviceId': 'D00KSA300VJDWW',
  'platform': 'yandexmodule_2'
}
interface_index=None
host_ttl=120
other_ttl=4500
```
То есть кроме name/deviceId весь сетевой ServiceInfo у них фактически общий.

Для сравнения четвёртый Module 2:
```
deviceId='D00CSA300W1Y2W'
server='Android-3.local.'
addresses=['192.168.3.53']
parsed_addresses=[
  '192.168.3.53',
  'fe80::ba87:6eff:fe47:749c'
]
properties={
  'td': 'follower',
  'cluster': 'yes',
  'deviceId': 'D00CSA300W1Y2W',
  'platform': 'yandexmodule_2'
}
```
Он этой проблемы не имеет.

Ещё важный момент: для проблемного устройства реальный IP может не только менять позицию в addresses, но и временно вообще исчезать из списка. Например для D00KSA300VJDWW, реальный IP которого .52, были последовательно:

['192.168.3.52']
['192.168.3.54', '192.168.3.52']
['192.168.3.52', '192.168.3.54']
['192.168.3.55', '192.168.3.52', '192.168.3.54']
['192.168.3.55', '192.168.3.54']

Поэтому вариант «сохранять текущий host, пока он присутствует в info.addresses» уменьшит число reconnect, но полностью проблему не решит.

За ~18 минут handler вызвался для трёх устройств с Android.local. примерно по 70–75 раз для каждого, а для Module с Android-3.local. — всего 2 раза.

Если нужно, приложу полный debug-log.

**Comment by AlexxIT** [comment]
https://github.com/AlexxIT/YandexStation/releases/tag/v3.22.0

**Comment by AlexxIT** [comment]
Рекомендую указать все host вручную и отключить автоматическое локальное обнаружение
https://github.com/AlexxIT/YandexStation#%D1%81%D1%82%D0%B0%D1%82%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D0%B9-ip-%D0%B4%D0%BB%D1%8F-%D0%BA%D0%BE%D0%BB%D0%BE%D0%BD%D0%BA%D0%B8
